kbfsgit/runner.go

```diff
@@ -503,6 +514,196 @@ func (r *runner) processGogitStatus(
+			chrootFS, err := fs.Chroot(fi.Name())
+			if err != nil {
+				return 0, 0, err
+			}
+			totalSoFar, toErase, err = r.recursiveByteCount(
+				ctx, chrootFS, totalSoFar, toErase)
+			if err != nil {
+				return 0, 0, err
+			}
+		} else {
+			totalSoFar += fi.Size()
+			if r.progress {
+				// This function only runs if r.verbosity >= 1.
+				eraseStr := strings.Repeat("\b", toErase)
+				newStr := fmt.Sprintf("%d bytes... ", totalSoFar)
+				toErase = len(newStr)
+				r.errput.Write([]byte(eraseStr + newStr))
+			}
+		}
+	}
+	return totalSoFar, toErase, nil
+}
+
+// statusWriter is a simple io.Writer shim that logs to `r.errput` the
+// number of bytes written to `output`.
+type statusWriter struct {
+	r           *runner
+	output      io.Writer
+	soFar       int64
+	totalBytes  int64
+	nextToErase int
+}
+
+var _ io.Writer = (*statusWriter)(nil)
+
+func (sw *statusWriter) Write(p []byte) (n int, err error) {
+	n, err = sw.output.Write(p)
+	if err != nil {
+		return n, err
+	}
+
+	sw.soFar += int64(len(p))
+	eraseStr := strings.Repeat("\b", sw.nextToErase)
+	newStr := fmt.Sprintf("%d/%d bytes... ", sw.soFar, sw.totalBytes)
+	sw.r.errput.Write([]byte(eraseStr + newStr))
+	sw.nextToErase = len(newStr)
+	return n, nil
+}
+
+func (r *runner) copyFile(
+	ctx context.Context, fs billy.Filesystem, localFS billy.Filesystem,
+	name string, sw *statusWriter) (err error) {
+	f, err := fs.Open(name)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	localF, err := localFS.Create(name)
+	if err != nil {
+		return err
+	}
+	defer localF.Close()
+
+	var w io.Writer = localF
+	// Wrap the destination file in a status shim if we are supposed
+	// to report progress.
+	if sw != nil && r.progress {
+		sw.output = localF
+		w = sw
+	}
+	_, err = io.Copy(w, f)
+	return err
+}
+
+// recursiveCopy copies the entire subdirectory rooted at `fs` to
+// `localFS`.
+func (r *runner) recursiveCopy(
+	ctx context.Context, fs billy.Filesystem, localFS billy.Filesystem,
+	sw *statusWriter) (err error) {
+	fileInfos, err := fs.ReadDir("")
+	if err != nil {
+		return err
+	}
+
+	for _, fi := range fileInfos {
+		if fi.IsDir() {
+			err := localFS.MkdirAll(fi.Name(), 0775)
+			if err != nil {
+				return err
+			}
+			chrootFS, err := fs.Chroot(fi.Name())
+			if err != nil {
+				return err
+			}
+			chrootLocalFS, err := localFS.Chroot(fi.Name())
+			if err != nil {
+				return err
+			}
+			err = r.recursiveCopy(ctx, chrootFS, chrootLocalFS, sw)
+			if err != nil {
+				return err
+			}
+		} else {
+			err := r.copyFile(ctx, fs, localFS, fi.Name(), sw)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// handleClone copies all the object files of a KBFS repo directly
+// into the local git dir, instead of using go-git to calculate the
+// full set of objects that are to be transferred (which is slow and
+// memory inefficient).  If the user requested only a single branch of
+// cloning, this will copy more objects that necessary, but still only
```

Comment:
In the future we can follow this with a GC to remove the unnecessary ones, correct?

Reply:
Yes, will add a note.